Fixs v3.0 and new features

Erim32 opened the pull request:
Done:
- fix tweets, rt, comments, hashtags, dependencies
- fix follow update_tables
- group tweets search in the loop of users

Todo:
- Integrate bypass module
- Integrate unfollow
- Integrate stats
- Integrate Help module
- group in only one database

Commit:
Some fixs on actions management 

- fix tweets, rt, comments, hashtags, dependencies
- fix follow update_tables
- group tweets search in the loop of users

Commit:
Merge branch 'dev' into develop

j4rj4r commented on main.py:
```diff
@@ -68,16 +69,18 @@
+        # If there is no action
+        if not list_action:
+            logging.error('There is no action to do!')
+            sys.exit()
```
C'est pas mieux de rechercher les tweets avec un compte et de passer la liste aux autres après ? Il y a (un peu) moins de requete à l'api comme ca.

j4rj4r commented on main.py:
```diff
@@ -68,16 +69,18 @@
+        manage_follow = Manage_follow(user_information["user"], user_information["api"])
+
+        # We retrieve the list of actions to do
+        list_action = action.search_tweets(mainaccount)
```
C'est pas mieux de rechercher les tweets avec un compte et de passer la liste aux autres après ? Il y a (un peu) moins de requete à l'api comme ca.

Erim32 replied:
@j4rj4r  Si tu veux participer aux même concours oui, après j'ai appliqué cette modification car  il est possible qu'en fonction des préférences ( likes, follow précédents) ce ne sont pas les même tweets qui remontent ?
 Un peu comme des recommandations/suggestions, mais cela reste une hypothèse.

HitoTech replied:
Autre idée (qui prendra plus de temps) : vu qu'on sauvegarde les données des tweets en BDD pour ne pas participer plusieurs fois aux mêmes concours, est-ce qu'on ne ferait pas en sorte de sauvegarder plus d'infos sur ces tweets concours (genre l'url du concours), et faire une table de liaison qui permet de valider si un compte a deja participé ou pas ?

Comme ça, y'a un compte qui passe, qui sauvegarde cette liste, puis elle est réutilisée par tous les autres comptes.

Erim32 replied:
@HitoTech ça peut être tres rapide il faut juste une table
Giveaway ( ID, TWEET_ID, USER_ID)
dans laquelle on insère chaque participation d'utilisateurs
et on fait un giveaway_exists(newtweet.id, user.id) avec recherche sur les deux critères et on utilise en condition avant de participer.

C'est une bonne idée

Commit:
Integrate AntiBot bypass

Erim32 commented:
@j4rj4r  Ajout du bypass dans la PR

Commit:
New features and fix

- unfollow management
- log giveaways participations
- alerts system  ( email, telegram, discord, slack)
- detect and notify new private message
- log winning giveaway
- multi threading
- improve logging with user account
- able to run multiple account with participation and notification 
simultaneously.
- update config file with alerter configurations
- update readme with alerter usage
- improve constants management
- unify database model on one database file
- add header

Depreciated: Stats table

Commit:
fix : Sleep message, variable interpolation

Commit:
Optimize  unfollow & replace print to logging.info

j4rj4r commented on BotTwitter/manage_follow.py:
```diff
@@ -15,72 +17,55 @@ def __init__(self, user, api):
+        if len(follows_rows) > 0:
+            # already follow, update the date with the current date
+            self.database_follows.update_follow(FollowId=follows_rows[0].FollowId, 
+                                                new_UserId=str(self.user.screen_name), 
```
Yo, J'avais un doute mais en regardant la doc de la base de données,  j'ai vu que l'UserId correspond à l'id (qui s'auto incremente) à l'ajout dans la base de données de l'utilisateur (donc la correspondance screen_name -> Userid se trouve dans la table user) .

Autre petit détail mais qui va nous éviter des bugs ^^ Tu utilises le user.screen_name qui est le nom afficher sur twitter mais ce nom peut etre changé (si il est changé il peut y avoir un problème avec la bdd). Je pense que le mieux est d'utiliser le user.id_str qui est un id donné pour chaque utilisateur par Twitter et qui ne peut pas etre changé.

Erim32 replied:
Bien vu, il faut remplacer self.user.screen_name par self.user.id ligne 35
Tu n'as de réel besoin de faire un ID local par utilisateur vu que l'ID twitter est deja unique ( la table Users est facultative).

j4rj4r replied:
Oui c'est vrai faut je change la doc et que j'enlève l'ID local

Erim32 replied:
Nouveau commit réalisé avec la modification

j4rj4r replied:
Du coup c'est bon pour moi ^^ Si tu le veux bien j'accepte le PR et on fait les autres modifs qu'il y a à faire dans d'autres PR

Erim32 replied:
C'est bon pour moi aussi 👍 oui, ça sera plus simple.

j4rj4r commented on BotTwitter/manage_follow.py:
```diff
@@ -15,72 +17,55 @@ def __init__(self, user, api):
-        c = connection.cursor()
-        c.execute("""SELECT * FROM {tab};""".format(tab=self.user.screen_name))
-        data = c.fetchall()
+        follows_rows = self.database_follows.get_follows(UserId=str(self.user.screen_name))
```
Un autre exemple ici

Commit:
Fix manage_follow, user.id replace user.screen_name
